x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java

```diff
@@ -0,0 +1,167 @@
+    }
+
+    public Join replaceChildren(LogicalPlan left, LogicalPlan right) {
+        return new Join(source(), left, right, config);
+    }
+
+    @Override
+    public List<Attribute> output() {
+        if (lazyOutput == null) {
+            lazyOutput = computeOutput();
+        }
+        return lazyOutput;
+    }
+
+    private List<Attribute> computeOutput() {
+        List<Attribute> right = makeReference(right().output());
+        return switch (config.type()) {
+            case LEFT -> // right side becomes nullable
+                mergeOutput(left().output(), makeNullable(right));
+            case RIGHT -> // left side becomes nullable
+                mergeOutput(makeNullable(left().output()), right);
+            case FULL -> // both sides become nullable
+                mergeOutput(makeNullable(left().output()), makeNullable(right));
+            default -> // neither side becomes nullable
+                mergeOutput(left().output(), right);
+        };
+    }
+
+    /**
+     * Merge output fields, left hand side wins in name conflicts <strong>except</strong>
+     * for fields defined in {@link JoinConfig#matchFields()}.
+     */
+    List<Attribute> mergeOutput(List<? extends Attribute> lhs, List<? extends Attribute> rhs) {
+        List<Attribute> results = new ArrayList<>(lhs.size() + rhs.size());
+
+        for (Attribute a : lhs) {
+            if (rhs.contains(a) == false || config.matchFields().stream().anyMatch(m -> m.name().equals(a.name()))) {
+                results.add(a);
+            }
+        }
+        for (Attribute a : rhs) {
+            if (false == config.matchFields().stream().anyMatch(m -> m.name().equals(a.name()))) {
+                results.add(a);
+            }
+        }
+        return results;
+    }
+
+    /**
+     * Make fields references, so we don't check if they exist in the index.
+     * We do this for fields that we know don't come from the index.
+     * TODO we should rework stats so we don't have to do this
+     */
+    public static List<Attribute> makeReference(List<Attribute> output) {
+        List<Attribute> out = new ArrayList<>(output.size());
+        for (Attribute a : output) {
+            if (a.resolved()) {
+                out.add(new ReferenceAttribute(a.source(), a.name(), a.dataType(), a.qualifier(), a.nullable(), a.id(), a.synthetic()));
+            } else {
+                out.add(a);
+            }
+        }
+        return out;
+    }
+
+    public static List<Attribute> makeNullable(List<Attribute> output) {
+        List<Attribute> out = new ArrayList<>(output.size());
+        for (Attribute a : output) {
+            out.add(a.withNullability(Nullability.TRUE));
+        }
+        return out;
+    }
+
+    @Override
+    public boolean expressionsResolved() {
+        return config.expressionsResolved();
+    }
+
+    public boolean duplicatesResolved() {
+        return left().outputSet().intersect(right().outputSet()).isEmpty();
```

Comment:
Just wanted to point out: I think this uses name ids to compute the intersection, not names; so I expect this to return `true` essentially always.

Reply:
I'll have to go rereview this one.

Reply:
This one always seems to return `true`. They never intersect so far as I can tell.

Reply:
Then let's remove it.